node/core/provisioner/src/disputes/mod.rs

```diff
@@ -0,0 +1,454 @@
+// Copyright 2017-2022 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::{
+	error::{Error, GetOnchainDisputesError},
+	metrics, LOG_TARGET,
+};
+use futures::channel::oneshot;
+use polkadot_node_primitives::{CandidateVotes, DisputeStatus};
+use polkadot_node_subsystem::{messages::DisputeCoordinatorMessage, overseer, ActivatedLeaf};
+use polkadot_primitives::v2::{
+	supermajority_threshold, CandidateHash, DisputeState, DisputeStatement, DisputeStatementSet,
+	Hash, InvalidDisputeStatementKind, MultiDisputeStatementSet, SessionIndex,
+	ValidDisputeStatementKind, ValidatorIndex, ValidatorSignature,
+};
+use std::collections::HashMap;
+
+#[cfg(test)]
+mod tests;
+
+//! This module implements `select_disputes` function which selects disputes (and votes) which should
+//! be sent to the Runtime.
+//!
+//! # How the prioritisation works
+//!
+//! Generally speaking disputes can be described as:
+//!   * Active vs Concluded
+//!   * Known vs Unknown onchain
+//!   * Local vs Onchain
+//!   * Runtime knows about enough votes to conclude the dispute vs
+//!     Runtime needs more votes to conclude the dispute.
+//!
+//! Provisioner fetches all disputes from `dispute-coordinator` and separates them in five partitions
+//! (listed in order of importance) based on this criteria:
+//!
+//!   1. Active local disputes, also known onchain, without enough votes to conclude onchain.
+//!   2. Active local disputes, unknown onchain.
+//!   3. Active local disputes, also known onchain, with enough votes to conclude onchain.
+//!   4. Concluded local disputes, unknown onchain.
+//!   5. Concluded local disputes, known onchain.
+//!
+//! `disputes-coordinator` first sends all disputes in group 1, then if there is more space - group 2 and
+//! then group 3. Group 4 is supposed to be empty but it is sent nevertheless. Group 5 is completely dropped.
+//!
+//! Or a bit more visually:
+//! 	- Active
+//! 		- Known onchain
+//! 			- Not enough votes to conclude (PARTITION 1)
+//! 			- With enough votes to conclude	(PARTITION 3)
+//! 		- Unknown onchain (PARTITION 2)
+//! 	- Concluded
+//! 		- Known onchain (THESE ARE DROPPED)
+//! 		- Unknown onchain (PARTITION 4)
+//!
+//! # Votes selection
+//!
+//! Besides the prioritisation described above the votes in each partition are filtered too. Provisioner
```

Comment:
```suggestion
//! Besides the prioritization described above the votes in each partition are filtered too. Provisioner
```

AE is our target language for both API naming and docs, we should write that down in CONTRIBUTING.md - out of scope here